sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,432 @@
+            toColumn.appendNulls(batchSize)
+          } else {
+            field.dataType match {
+              case BooleanType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0) == 1
+                toColumn.appendBooleans(batchSize, data)
+
+              case ByteType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toByte
+                toColumn.appendBytes(batchSize, data)
+              case ShortType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toShort
+                toColumn.appendShorts(batchSize, data)
+              case IntegerType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+              case LongType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0)
+                toColumn.appendLongs(batchSize, data)
+
+              case DateType =>
+                val data = fromColumn.asInstanceOf[LongColumnVector].vector(0).toInt
+                toColumn.appendInts(batchSize, data)
+
+              case TimestampType =>
+                val data = fromColumn.asInstanceOf[TimestampColumnVector]
+                toColumn.appendLongs(batchSize, fromTimestampColumnVector(data, 0))
+
+              case FloatType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0).toFloat
+                toColumn.appendFloats(batchSize, data)
+              case DoubleType =>
+                val data = fromColumn.asInstanceOf[DoubleColumnVector].vector(0)
+                toColumn.appendDoubles(batchSize, data)
+
+              case StringType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+              case BinaryType =>
+                val data = fromColumn.asInstanceOf[BytesColumnVector]
+                var index = 0
+                while (index < batchSize) {
+                  toColumn.appendByteArray(data.vector(0), data.start(0), data.length(0))
+                  index += 1
+                }
+
+              case DecimalType.Fixed(precision, scale) =>
+                val d = fromColumn.asInstanceOf[DecimalColumnVector].vector(0)
+                appendDecimalWritable(toColumn, precision, scale, d)
+
+              case dt =>
+                throw new UnsupportedOperationException(s"Unsupported Data Type: $dt")
+            }
+          }
+        } else if (!field.nullable || fromColumn.noNulls) {
+          field.dataType match {
+            case BooleanType =>
+              val data = fromColumn.asInstanceOf[LongColumnVector].vector
+              var index = 0
+              while (index < batchSize) {
```

Comment:
Can we use `WritableColumn.appendBooleans()`?

Reply:
It's not. `data(index)` is changed at each iteration.

Reply:
I thought `appendBooleans()` can be used since `toColumn.appendLongs()` is used for `LongType`. I realized `data` is `Long` as [you pointed out](https://github.com/apache/spark/pull/19943/files/d30225993c2c674898ea43e64b555b1283ca8e0c#r158894581) .

Reply:
hmm, so the orc batch always return long as value? that seems very inefficient...

Reply:
anyone knows why ORC does this? cc @mmccline

Reply:
The reason is ORC only have LongColumnVector and DoubleColumnVector for implementation simplicity.
All integral types use LongColumnVector and real types(float/double) use DoubleColumnVector.